### ISSUE TITLE
Destructuring property ends up as any when using { [key]:x} = aObject

### ISSUE THREAD

**Issue by theduffmtl** [Microsoft/TypeScript#20725]
**TypeScript Version:**  2.7.0-dev.201xxxxx

**Code**

```ts
interface Model {
    '0': number
}

const key = '0';
const keyLiteral: '0' = '0';

const myModel = {
    '0': 9
}
const { [key]: iendupAsAny } = myModel;

const { [keyLiteral]: iEndupAnyAsWell } = myModel;

const { ['0']: meImWellTyped } = myModel;
```

**Expected behavior:**
I would expect the 3 fields to be typed as number or a typescript error since noImplicitAny is true

**Actual behavior:**

the first 2 end up as any

**Comment by Kovensky** [comment]
This looks related to, but not quite the same as https://github.com/Microsoft/TypeScript/issues/16789

**Comment by sandersn** [comment]
Fixed by #21070

**Comment by mhegazy** [comment]
The examples in the OP should be fixed in latest.
